src/main/java/org/truffleruby/language/TruffleBootNodes.java

```diff
@@ -209,7 +209,7 @@ private Source loadMainSourceSettingDollarZero(
                 throw new JavaException(e);
             }
 
-            getContext().getCoreLibrary().getGlobalVariables().put("$0", dollarZeroValue);
+            getContext().getCoreLibrary().getGlobalVariables().getStorage("$0").setValueInternal(dollarZeroValue);
```

Comment:
I think we should rename `put` to `register` or so to emphase it's only for declaring new global variables (the check is useful as during startup we don't want to blindly override). But that's for another PR.